File: solutions/0240_Search_a_2D_Matrix_II/iter.java
// ⭐️

class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix == null) return false;
        int r2 = matrix.length;
        if (r2 == 0) return false;
        int c2 = matrix[0].length;
        if (c2 == 0) return false;
        if (target < matrix[0][0] || target > matrix[r2-1][c2-1]) return false;
        int r1 = 0, c1 = c2 - 1;
        while (r1 < r2 && c1 >= 0) {
            if (matrix[r1][c1] == target) return true;
            else if (matrix[r1][c1] < target) {
                r1 += 1;
            } else {
                c1 -= 1;
            }
        }
        return false;
    }
}

File: solutions/0498_Diagonal_Traverse/iter.kt
class Solution {
    fun findDiagonalOrder(matrix: Array<IntArray>): IntArray {
        if (matrix.isEmpty() || matrix[0].isEmpty()) {
            return IntArray(0)
        }
        val m = matrix.size - 1
        val n = matrix[0].size - 1
        var r = 0
        var c = 0
        var ret = IntArray(0)
        for (l in 0..m + n + 1) {
            if (l % 2 == 0) {
                val steps = Math.min(r, n - c)
                for (s in 0..steps) {
                    ret += matrix[r - s][c + s]
                }
                r -= steps
                c += steps
                if (c < n) {
                    c += 1
                } else {
                    r += 1
                }
            } else {
                val steps = Math.min(m - r, c)
                for (s in 0..steps) {
                    ret += matrix[r + s][c - s]
                }
                r += steps
                c -= steps
                if (r < m) {
                    r += 1
                } else {
                    c += 1
                }
            }
        }
        return ret
    }
}
